test/test-plugin-loader.js

```diff
@@ -0,0 +1,317 @@
+/**
+ * Copyright 2017 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+var shimmer = require('shimmer');
+var Module = require('module');
+var assert = require('assert');
+var proxyquire = require('proxyquire');
+var path = require('path');
+
+// Save logs because in some cases we want to verify that something was logged.
+var logs = {
+  error: '',
+  warn: '',
+  info: ''
+};
+
+// Facilitates loading "fake" modules upon calling require().
+var fakeModules = {};
+
+// Adds module moduleName to the set of fake modules, using mock as the object
+// being "exported" by this module. In addition, providing version makes it
+// accessible by calling findModuleVersion.
+function addModuleMock(moduleName, version, mock) {
+  fakeModules[moduleName.replace('/', path.sep)] = {
+    exports: mock,
+    version: version
+  };
+}
+
+// This function creates an object with just enough properties to appear to the
+// plugin loader as the trace agent. It accepts the list of plugins that the
+// plugin loader reads.
+function createFakeAgent(plugins) {
+  function writeToLog(log, data) {
```

Comment:
I don't think this has to be done as part of this pull request but it would be a nice clean up thing to do in the future.